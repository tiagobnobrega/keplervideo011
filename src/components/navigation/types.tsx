/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */

import { DrawerScreenProps } from '@amzn/react-navigation__drawer/lib/typescript/src/types';
import { StackNavigationProp } from '@amzn/react-navigation__stack';
import { StackScreenProps } from '@amzn/react-navigation__stack/lib/typescript/src/types';
import { ImageSourcePropType, TouchableOpacity } from 'react-native';
import { TitleData } from '../../types/TitleData';

export enum Screens {
  DEFAULT_SCREEN = 'Home',
  APP_DRAWER = 'AppDrawer',
  HOME_SCREEN = 'Home',
  DETAILS_SCREEN = 'Details',
  PLAYER_SCREEN = 'Player',
  SETTINGS_SCREEN = 'Settings',
  SEARCH_SCREEN = 'Search',
  SEARCH_RESULTS_SCREEN = 'SearchResultsScreen',
  FEEDBACK_SCREEN = 'FeedBackScreen',
}

export enum DrawerType {
  PERMANENT = 'permanent',
  FRONT = 'front',
  SLIDE = 'slide',
  BACK = 'back',
}

export type AppStackParamList = {
  [Screens.APP_DRAWER]: undefined;
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.FEEDBACK_SCREEN]: undefined;
  [Screens.SEARCH_RESULTS_SCREEN]: {
    searchKeyword: string;
  };
  [Screens.DETAILS_SCREEN]: {
    data: TitleData;
    sendDataOnBack: (id: number | string) => void;
  };
  [Screens.PLAYER_SCREEN]: {
    data: TitleData;
    sendDataOnBack: () => void;
  };
};

export type AppStackScreenProps<T extends keyof AppStackParamList> =
  StackScreenProps<AppStackParamList, T>;

export type HomeScreenNavigationProps = StackNavigationProp<
  AppStackParamList,
  Screens.HOME_SCREEN
>;

export type AppDrawerParamList = {
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
};

export type AppDrawerScreenProps<T extends keyof AppDrawerParamList> =
  DrawerScreenProps<AppDrawerParamList, T>;

declare global {
  namespace ReactNavigation {
    interface RootParamList extends AppStackParamList {}
  }
}

export interface ButtonConfig {
  onPress: () => void;
  image: ImageSourcePropType;
  label: string;
  ref?: React.RefObject<TouchableOpacity>;
  testID: string;
}
