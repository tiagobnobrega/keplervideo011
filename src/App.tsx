/*
 * Copyright (c) 2022 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */
import { ThemeProvider } from '@amzn/kepler-ui-components';
import {
  useHideSplashScreenCallback,
  usePreventHideSplashScreen,
} from '@amzn/react-native-kepler';
import { NavigationContainer } from '@amzn/react-navigation__native';
import { Store } from '@reduxjs/toolkit';
import React, { useEffect, useMemo, useState } from 'react';
import { Provider } from 'react-redux';
import AppStack from './components/navigation/AppStack';
import { initializeStore } from './store';
import { createTheme } from './styles/ThemeBuilders';

const App = () => {
  usePreventHideSplashScreen();

  const [isAppReady, setIsAppReady] = useState(false);
  const [store, setStore] = useState<Store | null>(null);

  setTimeout(() => setIsAppReady(true), 4000);
  const hideSplashScreenCallback = useHideSplashScreenCallback();

  const initStore = async () => {
    const initializedStore = await initializeStore();
    setStore(initializedStore);
  };

  useEffect(() => {
    initStore();
  }, []);

  useEffect(() => {
    if (isAppReady) {
      // Hide the splash screen and render content
      hideSplashScreenCallback();
    }
  }, [isAppReady, hideSplashScreenCallback]);
  const theme = useMemo(() => createTheme(), []);

  if (!store) {
    return null;
  }

  return (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <NavigationContainer>
          <AppStack />
        </NavigationContainer>
      </ThemeProvider>
    </Provider>
  );
};

export default App;
