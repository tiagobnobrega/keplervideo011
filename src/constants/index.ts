import { COLORS } from '../styles/Colors';

export enum DPADEventType {
  LEFT = 'left',
  RIGHT = 'right',
  UP = 'up',
  DOWN = 'down',
  PLAY = 'play',
  PAUSE = 'pause',
  PLAYPAUSE = 'playpause',
  SKIPBACKWARD = 'skip_backward',
  SKIPFORWARD = 'skip_forward',
  SELECT = 'select',
  MENU = 'menu',
  BACK = 'back',
  PAGE_UP = 'page_up',
  PAGE_DOWN = 'page_down',
  PAGE_LEFT = 'page_left',
  PAGE_RIGHT = 'page_right',
  INFO = 'info',
  MORE = 'more',
}

export const NetworkStatus = {
  CONNECTED: 'Connected',
  NOT_CONNECTED: 'Not Connected',
};

export const LoginStatus = {
  SIGNED_IN: true,
  SIGNED_OUT: false,
};

export const EVENT_KEY_DOWN = 0;
export const EVENT_KEY_UP = 1;

export const MAX_NARROW_SCREEN_WIDTH = 1300;

export const CAPTION_DISABLE_TEXT = 'Off';
export const CAPTION_DISABLE_ID = '-1';
export const L2_GRADIENT_COLORS = [
  `${COLORS.PALE_GRAY}99`,
  `${COLORS.MEDIUM_GRAY}99`,
  COLORS.CHARCOAL_GRAY,
];
export enum ReadyState {
  HAVE_NOTHING = 0,
  HAVE_METADATA = 1,
  HAVE_CURRENT_DATA = 2,
  HAVE_FUTURE_DATA = 3,
  HAVE_ENOUGH_DATA = 4,
}
