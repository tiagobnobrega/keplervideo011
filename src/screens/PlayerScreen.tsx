import {
  ContentPersonalizationServer,
  PlaybackState,
} from '@amzn/kepler-content-personalization';
import {
  BackHandler,
  HWEvent,
  KeplerAppStateChange,
  useAddKeplerAppStateListenerCallback,
  useTVEventHandler,
} from '@amzn/react-native-kepler';
import { SafeAreaView } from '@amzn/react-native-safe-area-context';
import {
  KeplerCaptionsView,
  KeplerVideoSurfaceView,
  VideoPlayer,
} from '@amzn/react-native-w3cmedia';
import { isEqual } from 'lodash';
import React, { useCallback, useEffect, useRef } from 'react';

import { RouteProp } from '@amzn/react-navigation__core';
import { StackNavigationProp } from '@amzn/react-navigation__stack';
import {
  AppStateStatus,
  Platform,
  StyleSheet,
  TouchableOpacity,
  useWindowDimensions,
} from 'react-native';
import {
  AppStackParamList,
  AppStackScreenProps,
  Screens,
} from '../components/navigation/types';
import { isContentPersonalizationEnabled } from '../config/AppConfig';
import { CAPTION_DISABLE_ID, EVENT_KEY_DOWN, ReadyState } from '../constants';
import { getMockPlaybackEventForVideo } from '../mocks/ContentPersonalizationMocks';
import { COLORS } from '../styles/Colors';
import { VideoHandler } from '../utils/VideoHandler';
import BufferingWindow from '../w3cmedia/mediacontrols/BufferingWindow';
import VideoPlayerUI from '../w3cmedia/mediacontrols/VideoPlayerUI';
import { ShakaPlayer } from '../w3cmedia/shakaplayer/ShakaPlayer';

const BACKGROUND_STATE: AppStateStatus = 'background';

interface PlayerProps {
  navigation: StackNavigationProp<AppStackParamList, Screens.PLAYER_SCREEN>;
  route: RouteProp<AppStackParamList, Screens.PLAYER_SCREEN>;
}

const PlayerScreen = ({
  navigation,
  route,
}: AppStackScreenProps<Screens.PLAYER_SCREEN>) => {
  const { data, sendDataOnBack } = route.params;
  const { width: deviceWidth, height: deviceHeight } = useWindowDimensions();
  const addKeplerAppStateListenerCallback =
    useAddKeplerAppStateListenerCallback();
  const [showBuffering, setShowBuffering] = React.useState<boolean>(true);
  // After media initialization once media loads isVideoInitialized will change to true.
  const [isVideoInitialized, setIsVideoInitialized] =
    React.useState<boolean>(false);
  const surfaceHandle = useRef<string | null>(null);
  const captionViewHandle = useRef<string | null>(null);

  const captionStatus = useRef<boolean | null>(null);
  const setCaptionStatus = (status: boolean) => {
    captionStatus.current = status;
  };
  const [captionID, setCaptionID] = React.useState<string>(CAPTION_DISABLE_ID);
  const setSelectedCaptionInVideoPlayer = (id: string) => {
    setCaptionID(id);
  };
  const [isVideoEnded, setVideoEnded] = React.useState<boolean>(false);
  const timer = useRef<null | ReturnType<typeof setTimeout> | number>(null);

  const videoRef = useRef<VideoPlayer | null>(null);
  const player = useRef<ShakaPlayer | null>(null);
  const TIME_TO_GO_BACK_IF_VIDEO_ENDS = 2300;
  const BACK_NAVIGATION_DELAY = 700;

  const videoHandler = new VideoHandler(
    videoRef,
    player,
    data,
    setIsVideoInitialized,
    setVideoEnded,
  );

  useTVEventHandler((evt: HWEvent) => {
    if (!Platform.isTV) {
      return;
    }
    if (evt && evt.eventKeyAction === EVENT_KEY_DOWN) {
      if (isVideoEnded) {
        clearTimeout(Number(timer.current));
        setVideoEnded(false);
      }
    }
  });

  useEffect(() => {
    if (!isVideoEnded) {
      return;
    }
    timer.current = setTimeout(() => {
      navigateBack();
    }, TIME_TO_GO_BACK_IF_VIDEO_ENDS);

    return () => clearTimeout(Number(timer.current));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isVideoEnded]);

  useEffect(() => {
    //Initialize media player
    videoHandler.preBufferVideo();
    setupEventListeners();
    const changeSubscription = addKeplerAppStateListenerCallback(
      'change',
      handleAppStateChange,
    );
    return () => {
      changeSubscription.remove();
      clearTimeout(Number(timer.current));
      removeEventListeners();
      sendDataOnBack();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const handleAppStateChange = (nextAppState: KeplerAppStateChange): void => {
    if (!videoRef.current || !videoRef?.current?.currentTime) {
      return;
    }
    if (
      nextAppState === BACKGROUND_STATE &&
      videoRef.current.currentTime > 0 &&
      !videoRef.current.paused &&
      !videoRef.current.ended &&
      videoRef.current.readyState > ReadyState.HAVE_CURRENT_DATA
    ) {
      setShowBuffering(false);
      videoRef.current?.pause();
      reportVideoPaused();
    }
  };

  const reportVideoPaused = useCallback(() => {
    if (!isContentPersonalizationEnabled()) {
      return;
    }

    console.log(
      'k_content_per: Reporting new playback event while navigating back',
    );
    try {
      const playbackEvent = getMockPlaybackEventForVideo(
        videoRef,
        videoRef.current!.currentSrc,
        PlaybackState.PAUSED,
      );
      ContentPersonalizationServer.reportNewPlaybackEvent(playbackEvent);
    } catch (e) {
      console.error(`k_content_per: ${e}`);
    }
  }, []);

  useEffect(() => {
    return () => {
      /*
        When the player is unmounted we check if the instance is correctly destroyed,
        if not, it is destroyed (Used commonly with integration like EPG)
      */
      if (videoRef?.current) {
        surfaceHandle.current = null;
        captionViewHandle.current = null;
        videoRef.current?.clearSurfaceHandle('');
        videoRef?.current?.clearCaptionViewHandle('');
        videoHandler.destroyVideoElements();
        videoRef.current = null;
        clearTimeout(Number(timer.current));
      }
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const navigateBack = useCallback(() => {
    if (captionStatus.current) {
      captionStatus.current = false;
      return true;
    }
    surfaceHandle.current = null;
    captionViewHandle.current = null;
    videoRef?.current?.clearSurfaceHandle('');
    videoRef?.current?.clearCaptionViewHandle('');
    videoHandler.destroyVideoElements();
    videoRef.current = null;
    clearTimeout(Number(timer.current));

    //Approx delay added to destroy media instance
    setTimeout(() => navigation.goBack(), BACK_NAVIGATION_DELAY);
    return true;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const setupEventListeners = useCallback(() => {
    if (Platform.isTV) {
      BackHandler.addEventListener('hardwareBackPress', navigateBack);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [Platform.isTV]);

  const removeEventListeners = useCallback(() => {
    if (Platform.isTV) {
      BackHandler.removeEventListener('hardwareBackPress', navigateBack);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [Platform.isTV]);

  const reportVideoPlaying = useCallback(() => {
    if (!isContentPersonalizationEnabled()) {
      return;
    }

    console.info('k_content_per: Reporting new playback event on playVideo()');
    try {
      const playbackEvent = getMockPlaybackEventForVideo(
        videoRef,
        videoRef.current!.currentSrc,
        PlaybackState.PLAYING,
      );
      console.info(
        `k_content_per: Reporting new playback event for PLAYING state: ${JSON.stringify(
          playbackEvent,
        )}`,
      );
      ContentPersonalizationServer.reportNewPlaybackEvent(playbackEvent);
    } catch (e) {
      console.error(`k_content_per: ${e}`);
    }
  }, [videoRef]);

  const playVideo = useCallback(() => {
    // Hide bufferview and Play the video if allowed...
    videoRef?.current?.play();
    setShowBuffering(false);
    reportVideoPlaying();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [videoRef]);

  const setSurface = useCallback(() => {
    if (surfaceHandle.current) {
      videoRef.current?.setSurfaceHandle(surfaceHandle.current);
      // Play video once surface has created and set to media reference
      playVideo();
    }
  }, [playVideo]);

  const onSurfaceViewCreated = useCallback(
    (_surfaceHandle: string): void => {
      console.info('Creating video surface');
      surfaceHandle.current = _surfaceHandle;
      setSurface();
    },
    [setSurface],
  );

  const onSurfaceViewDestroyed = useCallback((_surfaceHandle: string): void => {
    // Clear surface on screen unmount
    videoRef.current?.clearSurfaceHandle(_surfaceHandle);
  }, []);

  const onCaptionViewCreated = useCallback((captionsHandle: string): void => {
    console.info('Creating video captions');
    captionViewHandle.current = captionsHandle;
    setCaptionViewHandle();
  }, []);

  const setCaptionViewHandle = () => {
    console.info('setting captionViewHandle on video');
    if (videoRef.current !== null && captionViewHandle.current !== null) {
      (videoRef.current as VideoPlayer).setCaptionViewHandle(
        captionViewHandle.current,
      );
    }
  };
  const onCaptionViewDestroyed = (captionsHandle: string): void => {
    console.log(`Destroying captionsHandle=${captionsHandle}`);
    if (videoRef.current !== null) {
      (videoRef.current as VideoPlayer).clearCaptionViewHandle(captionsHandle);
    }
    captionViewHandle.current = null;
  };

  return (
    <SafeAreaView style={styles.playerContainer} testID="safe-area-view">
      <TouchableOpacity
        activeOpacity={1}
        style={[
          styles.surfaceContainer,
          {
            width: deviceWidth,
            height: deviceHeight,
          },
        ]}
        testID="touchable-opacity">
        {isVideoInitialized && (
          <>
            <KeplerVideoSurfaceView
              style={styles.videoSurface}
              onSurfaceViewCreated={onSurfaceViewCreated}
              onSurfaceViewDestroyed={onSurfaceViewDestroyed}
              testID="kepler-video-surface-view"
            />
            <KeplerCaptionsView
              onCaptionViewCreated={onCaptionViewCreated}
              onCaptionViewDestroyed={onCaptionViewDestroyed}
              style={styles.captions}
              show={captionID !== CAPTION_DISABLE_ID}
              testID="kepler-captions-view"
            />
          </>
        )}
        {showBuffering && <BufferingWindow />}
        {!showBuffering && (
          <VideoPlayerUI
            videoRef={videoRef}
            navigateBack={navigateBack}
            title={route.params.data.title}
            setCaptionStatus={setCaptionStatus}
            setSelectedCaptionInVideoPlayer={setSelectedCaptionInVideoPlayer}
          />
        )}
      </TouchableOpacity>
    </SafeAreaView>
  );
};

const playerScreenUnchanged = (
  prevProps: PlayerProps,
  nextProps: PlayerProps,
) => {
  return isEqual(prevProps, nextProps);
};
export default React.memo(PlayerScreen, playerScreenUnchanged);

export const styles = StyleSheet.create({
  captions: {
    width: '100%',
    height: '100%',
    top: 0,
    left: 0,
    position: 'absolute',
    backgroundColor: 'transparent',
    flexDirection: 'column',
    alignItems: 'center',
    zIndex: 2,
  },
  playerContainer: { backgroundColor: COLORS.GRAY, height: '100%' },
  surfaceContainer: { backgroundColor: COLORS.WHITE, alignItems: 'stretch' },

  videoSurface: { zIndex: 0 },
});
