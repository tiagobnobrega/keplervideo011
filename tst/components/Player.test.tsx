import { SafeAreaView } from '@amzn/react-native-safe-area-context';
import { VideoPlayer } from '@amzn/react-native-w3cmedia';
import { RouteProp } from '@amzn/react-navigation__core';
import { StackNavigationProp } from '@amzn/react-navigation__stack';
import { describe } from '@jest/globals';
import '@testing-library/jest-native/extend-expect';
import { fireEvent, render } from '@testing-library/react-native';
import { default as React } from 'react';
import 'react-native';
import { BackHandler } from 'react-native';
import {
  AppStackParamList,
  Screens,
} from '../../src/components/navigation/types';
import { CAPTION_DISABLE_ID } from '../../src/constants';
import PlayerScreen, { styles } from '../../src/screens/PlayerScreen';
import { TitleData } from '../../src/types/TitleData';
import { VideoHandler } from '../../src/utils/VideoHandler';
import { ShakaPlayer } from '../../src/w3cmedia/shakaplayer/ShakaPlayer';

const tileData: TitleData = {
  id: '169313',
  title: 'Beautiful Whale Tail Uvita Costa Rica',
  description: 'Beautiful Whale Tail Uvita Costa Rica',
  duration: 86,
  thumbnail:
    'http://le2.cdn01.net/videos/0000169/0169313/thumbs/0169313__007f.jpg',
  posterUrl:
    'http://le2.cdn01.net/videos/0000169/0169313/thumbs/0169313__007f.jpg',
  videoUrl:
    'https://edge-vod-media.cdn01.net/encoded/0000169/0169313/video_1880k/T7J66Z106.mp4?source=firetv&channelID=13454',
  categories: ['Costa Rica Islands'],
  channelID: '13454',
  rating: '2.5',
  mediaType: 'video',
  mediaSourceType: 'url',
  format: 'MP4',
  secure: false,
  uhd: true,
  rentAmount: '135',
} as TitleData;

const tileDataTwo: TitleData = {
  id: '169322',
  title: 'Big Buck Bunny: The Dark Truths',
  description: 'HLS format Big Buck Bunny: The Dark Truths',
  duration: 205,
  thumbnail:
    'http://le2.cdn01.net/videos/0000169/0169313/thumbs/0169313__007f.jpg',
  posterUrl:
    'http://le2.cdn01.net/videos/0000169/0169313/thumbs/0169313__007f.jpg',
  videoUrl:
    'https://storage.googleapis.com/shaka-demo-assets/bbb-dark-truths-hls/hls.m3u8',
  categories: ['Costa Rica Underwater'],
  channelID: '13455',
  rating: '2.5',
  mediaType: 'video',
  mediaSourceType: 'url',
  format: 'HLS',
  secure: false,
  uhd: false,
  textTrack: [
    {
      label: 'Sample CC 1',
      language: 'en',
      uri: 'https://mtoczko.github.io/hls-test-streams/test-vtt-ts-segments/text/1.vtt',
      mimeType: 'application/x-subtitle-vtt',
    },
    {
      label: 'Sample CC 2',
      language: 'es',
      uri: 'https://brenopolanski.github.io/html5-video-webvtt-example/MIB2-subtitles-pt-BR.vtt',
      mimeType: 'application/x-subtitle-vtt',
    },
  ],
  rentAmount: '135',
} as TitleData;

jest.mock('react-native', () => ({
  ...jest.requireActual('react-native'),
  isTV: true,
}));
jest.mock('react-native', () => ({
  BackHandler: {
    addEventListener: jest.fn(() => ({
      remove: jest.fn(),
    })),
    removeEventListener: jest.fn(),
  },
}));
jest.mock('../../src/constants', () => ({
  ReadyState: jest.requireActual('../../src/constants'),
  CAPTION_DISABLE_ID: jest.requireActual('../../src/constants')
    .CAPTION_DISABLE_ID,
}));

jest.mock('../../src/utils/VideoHandler', () => {
  let internalData = tileData;
  let internalFileType = tileData.format;
  const vidRef: React.MutableRefObject<VideoPlayer | null> = {
    current: {
      setSurfaceHandle: jest.fn(),
      setCaptionViewHandle: jest.fn(),
      currentTime: 1000,
      readyState: 3,
      play: jest.fn(),
      pause: jest.fn(),
      clearSurfaceHandle: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
    } as unknown as VideoPlayer,
  };
  const playerRef: React.MutableRefObject<ShakaPlayer | null> = {
    current: {
      load: jest.fn(),
      unload: jest.fn(),
    } as unknown as ShakaPlayer,
  };
  return {
    VideoHandler: jest.fn().mockImplementation(() => ({
      selectedFileType: 'MP4',
      videRef: vidRef,
      playerRef: playerRef,
      data: internalData,
      internalFileType: internalFileType,
      destroyVideoElements: jest.fn(),
      preBufferVideo: jest.fn(),
      onLoadeMetaData: jest.fn(),
    })),
  };
});

jest.mock('@amzn/react-native-safe-area-context', () => {
  const MockSafeAreaView = ({ children }: { children: React.ReactNode }) => (
    <>{children}</>
  );

  return {
    __esModule: true,
    SafeAreaView: MockSafeAreaView,
  };
});

jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');
  RN.Dimensions.get = jest.fn().mockImplementation(() => {
    return {
      width: 375,
      height: 667,
    };
  });
  return RN;
});

const videoRef: React.MutableRefObject<VideoPlayer | null> = {
  current: {
    setSurfaceHandle: jest.fn(),
    setCaptionViewHandle: jest.fn(),
    currentTime: 1000,
    readyState: 3,
    play: jest.fn(),
    pause: jest.fn(),
    clearSurfaceHandle: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
  } as unknown as VideoPlayer,
};

const playerRef: React.MutableRefObject<ShakaPlayer | null> = {
  current: {
    load: jest.fn(),
    unload: jest.fn(),
  } as unknown as ShakaPlayer,
};

const mockedNavigate = jest.fn();
const sendDataOnBack = jest.fn();
const mockedNavigation = {
  navigate: mockedNavigate,
  goBack: mockedNavigate,
} as unknown as StackNavigationProp<
  AppStackParamList,
  Screens.PLAYER_SCREEN,
  undefined
>;
const mockRoute = {
  key: '',
  params: {
    data: tileData,
    sendDataOnBack: sendDataOnBack,
  },
  name: Screens.PLAYER_SCREEN,
} as unknown as RouteProp<AppStackParamList, Screens.PLAYER_SCREEN>;

const renderPlayerScreen = () => {
  return (
    <PlayerScreen
      navigation={mockedNavigation}
      route={{
        ...mockRoute,
        params: {
          ...mockRoute.params,
        },
      }}
    />
  );
};

jest.mock('../../src/w3cmedia/mediacontrols/BufferingWindow');
jest.mock('../../src/w3cmedia/mediacontrols/VideoPlayerUI');

describe('PlayerScreen rendering Test Cases', () => {
  const setState = jest.fn();
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  let mockSetState: any;
  beforeEach(() => {
    jest.clearAllMocks();
    mockSetState = jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => [true, setState]) //setShowBuffering
      .mockImplementationOnce(() => [false, setState]) //setIsVideoInitialized
      .mockImplementationOnce(() => [CAPTION_DISABLE_ID, setState]); //setCaptionID
  });

  it('renders correctly screenshot match', () => {
    const tree = render(renderPlayerScreen());
    expect(tree).toMatchSnapshot();
  });

  it('SafeAreaView is present in component', async () => {
    const { queryByTestId } = render(renderPlayerScreen());
    const safeAreaView = queryByTestId('safe-area-view');
    expect(safeAreaView).toBeDefined();
  });

  it('SafeAreaView is rendered with correct styles', async () => {
    const { UNSAFE_queryAllByType } = render(renderPlayerScreen());
    const safeAreaView = UNSAFE_queryAllByType(SafeAreaView);
    const safeAreaViewWithStyles = safeAreaView.find(
      safeArea => safeArea.props.style === styles.playerContainer,
    );
    expect(safeAreaViewWithStyles).toBeTruthy();
  });

  it('TouchableOpacity is present in component', async () => {
    const { queryByTestId } = render(renderPlayerScreen());
    const touchableOpacity = queryByTestId('touchable-opacity');
    expect(touchableOpacity).toBeDefined();
  });

  it('KeplerVideoSurfaceView is present in component', async () => {
    const { queryByTestId } = render(renderPlayerScreen());
    const keplerVideoSurfaceView = queryByTestId('kepler-video-surface-view');
    expect(keplerVideoSurfaceView).toBeDefined();
  });

  it('KeplerVideoSurfaceView should not show on initial render', async () => {
    const { queryByTestId } = render(renderPlayerScreen());
    const keplerVideoSurfaceView = queryByTestId('kepler-video-surface-view');
    expect(keplerVideoSurfaceView).not.toBeOnTheScreen();
  });

  it('KeplerCaptionsView is present in component', async () => {
    const { queryByTestId } = render(renderPlayerScreen());
    const keplerCaptionsView = queryByTestId('kepler-captions-view');
    expect(keplerCaptionsView).toBeDefined();
  });

  it('KeplerCaptionsView should not show on initial render', async () => {
    const { queryByTestId } = render(renderPlayerScreen());
    const keplerCaptionsView = queryByTestId('kepler-captions-view');
    expect(keplerCaptionsView).not.toBeOnTheScreen();
  });

  it('BufferingWindow is present in component', async () => {
    const { queryByTestId } = render(renderPlayerScreen());
    const bufferingWindow = queryByTestId('buffering-view');
    expect(bufferingWindow).toBeDefined();
  });

  it('BufferingWindow should show on initial render', async () => {
    const { queryByTestId } = render(renderPlayerScreen());
    const bufferingWindow = queryByTestId('buffering-view');
    expect(bufferingWindow).not.toBeOnTheScreen();
  });

  it('VideoPlayerUI is present in component', async () => {
    const { queryByTestId } = render(renderPlayerScreen());
    const videoPlayerUIView = queryByTestId('video-player-ui-view');
    expect(videoPlayerUIView).toBeDefined();
  });

  it('VideoPlayerUi should not show on initial render', async () => {
    const { queryByTestId } = render(renderPlayerScreen());
    const videoPlayerUi = queryByTestId('video-player-ui-view');
    expect(videoPlayerUi).not.toBeOnTheScreen();
  });
});
describe('VideoHandler.....', () => {
  let setIsVideoInitialized, setIsVideoEnded;
  beforeEach(() => {
    setIsVideoInitialized = jest.fn();
    setIsVideoEnded = jest.fn();
  });
  test('should initialize VideoHandler with correct values', () => {
    const handler = new VideoHandler(
      videoRef,
      playerRef,
      tileData,
      setIsVideoInitialized,
      setIsVideoEnded,
    );
    expect(handler).not.toBeNull();
    expect(handler.selectedFileType).toEqual('MP4');
    expect(handler.videoRef).not.toBeNull();
    expect(handler.player).not.toBeNull();
    expect(handler.data).toBe(tileData);
  });
});

describe('PlayerScreen with React.memo', () => {
  it('renders only when props change', async () => {
    const { rerender, getByTestId } = render(
      <PlayerScreen navigation={mockedNavigation} route={mockRoute} />,
    );
    const isEqualSpy = jest.spyOn(require('lodash'), 'isEqual');
    rerender(<PlayerScreen navigation={mockedNavigation} route={mockRoute} />);
    expect(isEqualSpy).toHaveBeenCalledWith(
      {
        navigation: mockedNavigation,
        route: mockRoute,
      },
      {
        navigation: mockedNavigation,
        route: mockRoute,
      },
    );
    rerender(
      <PlayerScreen
        navigation={mockedNavigation}
        route={{
          ...mockRoute,
          params: {
            ...mockRoute.params,
            data: tileDataTwo,
          },
        }}
      />,
    );
    expect(isEqualSpy).toHaveBeenCalledWith(
      {
        navigation: mockedNavigation,
        route: mockRoute,
      },
      {
        navigation: mockedNavigation,
        route: {
          ...mockRoute,
          params: {
            ...mockRoute.params,
            data: tileDataTwo,
          },
        },
      },
    );
    expect(getByTestId('touchable-opacity')).toBeTruthy();
  });
});

describe('PlayerScreen hooks', () => {
  let component: any;
  let clearTimeOutSpy: any;

  beforeEach(() => {
    component = render(renderPlayerScreen());
    clearTimeOutSpy = jest.spyOn(global, 'clearTimeout');
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test('useTVEventHandler: clearTimeOut when video is finished', async () => {
    fireEvent(component.queryByTestId('touchable-opacity'), 'TVRemoteCommand', {
      eventKeyAction: 0,
    });
    expect(clearTimeOutSpy).not.toHaveBeenCalled();
  });

  test('BackHandler.addEventListener needs to defined', () => {
    jest
      .spyOn(BackHandler, 'addEventListener')
      .mockImplementation((eventName: any, navigateBack: any) => {
        if (eventName === 'hardwareBackPress') {
          navigateBack();
        }
        return {
          remove: () => {},
        };
      });

    expect(BackHandler.addEventListener).toBeDefined();
  });

  let mockAddKeplerAppStateListenerCallback;

  test('useAddKeplerAppStateListenerCallback ', () => {
    const {
      useAddKeplerAppStateListenerCallback,
    } = require('@amzn/react-native-kepler');
    const mockChangeSubscription = {
      remove: jest.fn(),
    };
    mockAddKeplerAppStateListenerCallback = jest
      .fn()
      .mockReturnValue(mockChangeSubscription);
    useAddKeplerAppStateListenerCallback.mockReturnValue(
      mockAddKeplerAppStateListenerCallback,
    );
    const { unmount } = render(renderPlayerScreen());
    expect(mockAddKeplerAppStateListenerCallback).toHaveBeenCalledWith(
      'change',
      expect.any(Function),
    );
    const handleAppStateChange =
      mockAddKeplerAppStateListenerCallback.mock.calls[0][1];
    handleAppStateChange('background');
    expect(mockChangeSubscription.remove).not.toHaveBeenCalled();
    unmount();
    expect(mockChangeSubscription.remove).toHaveBeenCalledWith();
  });

  test('useAddKeplerAppStateListenerCallback With videoRef.current as null', () => {
    const {
      useAddKeplerAppStateListenerCallback,
    } = require('@amzn/react-native-kepler');
    const mockChangeSubscription = {
      remove: jest.fn(),
    };
    mockAddKeplerAppStateListenerCallback = jest
      .fn()
      .mockReturnValue(mockChangeSubscription);
    useAddKeplerAppStateListenerCallback.mockReturnValue(
      mockAddKeplerAppStateListenerCallback,
    );
    const { unmount } = render(renderPlayerScreen());
    expect(mockAddKeplerAppStateListenerCallback).toHaveBeenCalledWith(
      'change',
      expect.any(Function),
    );
    const handleAppStateChange =
      mockAddKeplerAppStateListenerCallback.mock.calls[0][1];
    handleAppStateChange('background');
    expect(mockChangeSubscription.remove).not.toHaveBeenCalled();
    unmount();
    expect(mockChangeSubscription.remove).toHaveBeenCalledWith();
  });
});
