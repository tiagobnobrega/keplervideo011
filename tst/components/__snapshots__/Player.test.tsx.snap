// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerScreen rendering Test Cases renders correctly screenshot match 1`] = `
<View
  accessibilityState={
    Object {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    Object {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "stretch",
      "backgroundColor": "#FFFFFF",
      "height": undefined,
      "opacity": 1,
      "width": undefined,
    }
  }
  testID="touchable-opacity"
/>
`;
